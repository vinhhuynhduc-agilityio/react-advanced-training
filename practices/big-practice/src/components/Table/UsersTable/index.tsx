import React, {
  useEffect,
  useRef,
} from 'react';

// ag-grid
import {
  ColDef,
  GetRowIdParams,
  GridApi,
  GridReadyEvent,
  RowClassParams,
  RowClickedEvent,
  RowDoubleClickedEvent
} from 'ag-grid-community';
import 'ag-grid-community/styles/ag-grid.css';
import 'ag-grid-community/styles/ag-theme-alpine.css';

// styles
import '../../../style.css';

// types
import { UserData } from '@/types';

// component
import { Spinner } from '@/components/common';
import { DataGrid } from '@/components';
import { PersonListItem } from '@/components/DataGrid';

// hooks
import { useDashboardContext } from '@/hooks';

interface UsersTableProps {
  selectedUserId: string | null;
  sourceComponent: string | null;
  onUserSelected: (userId: string | null) => void;
  registerGridApi: (api: GridApi) => void;
  onUserDoubleClicked: (data: UserData) => void;
  isLoading: boolean;
  isSavingUser: boolean;
}

const UsersTable: React.FC<UsersTableProps> = ({
  selectedUserId,
  onUserSelected,
  sourceComponent,
  registerGridApi,
  onUserDoubleClicked,
  isLoading,
}) => {
  const { users } = useDashboardContext();
  const gridApi = useRef<GridApi | null>(null);

  const onGridReady = (params: GridReadyEvent) => {
    gridApi.current = params.api;

    // Pass GridApi to Dashboard via registerGridApi
    registerGridApi(params.api);
  }

  // When `selectedUserId` changes, scroll to the corresponding row
  useEffect(() => {
    if (
      sourceComponent !== 'UsersTable'
      && selectedUserId
      && gridApi.current
    ) {
      const rowNode = gridApi.current.getRowNode(selectedUserId);
      if (rowNode) {

        // Scroll to row with matching userId
        gridApi.current.ensureNodeVisible(rowNode, 'middle');
      }
    }
  }, [selectedUserId, sourceComponent]);

  const onRowClicked = (event: RowClickedEvent) => {
    const userId = event.data.id;

    if (userId) {

      // Pass `userId` to parent component
      onUserSelected(userId);
    }
  };

  const getRowClass = (params: RowClassParams) => {
    return params.data.id === selectedUserId
      ? 'ag-row-selected'
      : '';
  };

  const getRowId = (params: GetRowIdParams<UserData>): string => {
    return params.data.id;
  };

  const onRowDoubleClicked = (event: RowDoubleClickedEvent) => {
    onUserDoubleClicked(event.data);
  };

  const columnDefs: ColDef<UserData>[] = [
    {
      headerName: 'Persons',
      field: 'fullName',
      cellRenderer: PersonListItem,
      headerClass: 'custom-header',
      flex: 1
    },
  ];

  return (
    <DataGrid
      rowData={users}
      columnDefs={columnDefs}
      onRowClicked={onRowClicked}
      getRowClass={getRowClass}
      rowHeight={75}
      onGridReady={onGridReady}
      getRowId={getRowId}
      onRowDoubleClicked={onRowDoubleClicked}
      loadingOverlayComponent={Spinner}
      loading={isLoading}
    />
  );
};

export default UsersTable;
