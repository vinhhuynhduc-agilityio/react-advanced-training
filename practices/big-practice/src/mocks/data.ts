export const mockUsers = [
  {
    id: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    fullName: "Joe Bloggs",
    earnings: "$11500",
    email: "john@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=1",
    registered: "May 21, 2020 17:02:06",
    lastUpdated: "Nov 11, 2024 14:34:21",
  },
  {
    id: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    fullName: "Jane Smith",
    earnings: "$36911",
    email: "jane@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=2",
    registered: "June 10, 2021 09:20:15",
    lastUpdated: "Nov 18, 2024 17:33:37",
  },
  {
    id: "123e4567-e89b-12d3-a456-426614174000",
    fullName: "Alice Brown",
    earnings: "$7200",
    email: "alice@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=3",
    registered: "April 15, 2020 14:02:06",
    lastUpdated: "October 11, 2023 10:00:00",
  },
];

export const mockProject = [
  {
    id: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    projectName: "Support",
  },
  {
    id: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    projectName: "Failure Testing",
  },
  {
    id: "e38e56a2-4d3c-469d-8345-45d6b5fae9f9",
    projectName: "Quality Management",
  },
  {
    id: "f49b68f4-5e5b-457d-bf16-1b5b78e6f5f1",
    projectName: "Data Quality",
  },
  {
    id: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    projectName: "Test",
  },
];

export const mockTasks = [
  {
    id: "71e564f4-7c18-47f7-89f2-abe4b7ec2854",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    taskName: "Build test",
    startDate: "10 Aug 24",
    completedDate: "15 Nov 24",
    currency: 2000,
    status: true,
    projectName: "Support",
    fullName: "Joe Bloggs",
  },
  {
    id: "82e564f4-7c18-47f7-89f2-abe4b7ec2856",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Build Frontend",
    startDate: "05 Sep 23",
    completedDate: "27 Oct 24",
    currency: 4000,
    status: true,
    projectName: "Failure Testing",
    fullName: "Joe Bloggs",
  },
  {
    id: "u2e564f4-7c18-47f7-89f2-abe4b7ec2878",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Develop Chatbot Feature",
    startDate: "23 Sep 23",
    completedDate: "24 Oct 24",
    currency: 3200,
    status: true,
    projectName: "Failure Testing",
    fullName: "Jane Smith",
  },
  {
    id: "7ce698fc-c0ee-4d8c-8a1b-df988a96c22f",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "test today",
    startDate: "13 Nov 24",
    completedDate: "incomplete",
    currency: 1111,
    status: false,
    projectName: "Test",
    fullName: "Jane Smith",
  },
  {
    id: "r0s1t2u3-7c18-47f7-89f2-abe4b7ec2949",
    userId: "123e4567-e89b-12d3-a456-426614174000",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "Data Migration",
    startDate: "15 Aug 24",
    completedDate: "20 Feb 24",
    currency: 2800,
    status: true,
    projectName: "Test",
    fullName: "Alice Brown",
  },
];
