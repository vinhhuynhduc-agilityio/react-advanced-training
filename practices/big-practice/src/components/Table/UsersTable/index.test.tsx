import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import UsersTable from '.';
import { mockUsers } from '@/mocks';

describe('UsersTable component', () => {
  const defaultProps = {
    selectedUserId: null,
    onUserSelected: jest.fn(),
    sourceComponent: '',
    registerGridApi: jest.fn(),
    onUserDoubleClicked: jest.fn(),
    isLoading: false,
    isSavingUser: false,
    users: mockUsers,
  };

  const setup = (props = {}) => {
    const setupProps = { ...defaultProps, ...props };
    return render(<UsersTable {...setupProps} />);
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot for default state', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('renders user data correctly', () => {
    setup();

    // Check if the user names are rendered inside the ag-grid
    expect(screen.getByText('Joe Bloggs')).toBeInTheDocument();
    expect(screen.getByText('Jane Smith')).toBeInTheDocument();
  });

  it('renders the selected row with a special class', () => {
    setup({ selectedUserId: 'd290f1ee-6c54-4b01-90e6-d701748f0851' });

    // Check if the row for Joe Bloggs is selected
    expect(screen.getByText('Joe Bloggs').closest('.ag-row')).toHaveClass('ag-row-selected');
  });

  it('calls registerGridApi when the grid is ready', async () => {
    const mockRegisterGridApi = jest.fn();
    setup({ registerGridApi: mockRegisterGridApi });

    // Verify that the `registerGridApi` function was called
    await waitFor(() => expect(mockRegisterGridApi).toHaveBeenCalled());
  });

  it('calls onUserSelected when a row is clicked', async () => {
    const mockOnUserSelected = jest.fn();
    setup({ onUserSelected: mockOnUserSelected });

    // Simulate a click on the row for Joe Bloggs
    const row = screen.getByText('Joe Bloggs');
    if (row) {
      fireEvent.click(row);
    }

    // Check that the onUserSelected function was called with the correct user ID
    await waitFor(() => expect(mockOnUserSelected).toHaveBeenCalledWith('d290f1ee-6c54-4b01-90e6-d701748f0851'));
  });

  it('calls onUserDoubleClicked when a row is double-clicked', async () => {
    const mockOnUserDoubleClicked = jest.fn();
    setup({ onUserDoubleClicked: mockOnUserDoubleClicked });

    const rowElement = screen.getByText('Joe Bloggs').closest('.ag-row');

    if (!rowElement) {
      throw new Error('Row not found');
    }

    // Simulate a double-click event
    fireEvent.doubleClick(rowElement);

    // Check if the onUserDoubleClicked function was called with the correct user data
    await waitFor(() =>
      expect(mockOnUserDoubleClicked).toHaveBeenCalledWith(
        expect.objectContaining({
          id: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          fullName: 'Joe Bloggs',
          earnings: '$11500',
          email: 'john@example.com',
          avatarUrl: 'https://i.pravatar.cc/150?img=1',
          registered: 'May 21, 2020 17:02:06',
          lastUpdated: 'Nov 11, 2024 14:34:21',
        })
      )
    );
  });
});
