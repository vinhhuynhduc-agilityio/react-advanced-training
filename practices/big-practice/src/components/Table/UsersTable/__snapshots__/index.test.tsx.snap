// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable component matches snapshot for default state 1`] = `
<div>
  <div
    style="height: 100%; --ag-line-height: 75px;"
  >
    <div
      class="ag-measurement-container"
    >
      <div />
      <div />
      <div />
    </div>
    <div
      aria-atomic="true"
      aria-live="polite"
      aria-relevant="additions text"
      class="ag-aria-description-container"
    />
    <!-- AG Grid -->
    <div
      class="ag-root-wrapper ag-ltr ag-layout-normal"
      grid-id="1"
      role="presentation"
    >
      <div
        class="ag-root-wrapper-body ag-focus-managed ag-layout-normal"
        role="presentation"
      >
        <div
          class="ag-tab-guard ag-tab-guard-top"
          role="presentation"
          tabindex="0"
        />
        <!-- AG Grid Body -->
        <div
          aria-colcount="1"
          aria-rowcount="4"
          class="ag-root ag-unselectable ag-layout-normal ag-body-vertical-content-no-gap ag-body-horizontal-content-no-gap"
          role="treegrid"
        >
          <div
            class="ag-header ag-pivot-off ag-header-allow-overflow"
            role="presentation"
            style="height: 49px; min-height: 49px;"
          >
            <div
              aria-hidden="true"
              class="ag-pinned-left-header ag-hidden"
              role="rowgroup"
              style="width: 0px; min-width: 0px; max-width: 0px;"
            />
            <div
              class="ag-header-viewport "
              role="presentation"
            >
              <div
                class="ag-header-container"
                role="rowgroup"
                style="width: 200px;"
              >
                <div
                  aria-rowindex="1"
                  class="ag-header-row ag-header-row-column"
                  role="row"
                  style="height: 48px; top: 0px; width: 200px;"
                >
                  <div
                    aria-colindex="1"
                    aria-sort="none"
                    class="ag-header-cell ag-column-first ag-column-last ag-header-cell-sortable custom-header ag-focus-managed"
                    col-id="fullName"
                    role="columnheader"
                    style="top: 0px; height: 48px; width: 200px; left: 0px;"
                    tabindex="-1"
                  >
                    <div
                      aria-hidden="true"
                      class="ag-header-cell-resize ag-hidden"
                      role="presentation"
                    />
                    <div
                      aria-hidden="true"
                      class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                      role="presentation"
                    >
                      
                
                      <div
                        aria-hidden="true"
                        class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                        data-ref="eLabel"
                        id="ag-4-label"
                        role="presentation"
                      />
                      
                
                      <div
                        class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                        data-ref="eWrapper"
                        role="presentation"
                      >
                        
                    
                        <input
                          aria-label="Press Space to toggle all rows selection (unchecked)"
                          class="ag-input-field-input ag-checkbox-input"
                          data-ref="eInput"
                          id="ag-4-input"
                          tabindex="-1"
                          type="checkbox"
                        />
                        
                
                      </div>
                      
            
                    </div>
                    <div
                      class="ag-header-cell-comp-wrapper"
                      role="presentation"
                    >
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
        
                        
        
                        
        
                        <div
                          class="ag-header-cell-label"
                          data-ref="eLabel"
                          role="presentation"
                        >
                          
            
                          <span
                            class="ag-header-cell-text"
                            data-ref="eText"
                          >
                            Persons
                          </span>
                          
            
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            data-ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
            
                          <!--AG-SORT-INDICATOR-->
                          <span
                            class="ag-sort-indicator-container"
                            data-ref="eSortIndicator"
                          >
                            
        
                            <span
                              aria-hidden="true"
                              class="ag-sort-indicator-icon ag-sort-order ag-hidden"
                              data-ref="eSortOrder"
                            />
                            
        
                            <span
                              aria-hidden="true"
                              class="ag-sort-indicator-icon ag-sort-ascending-icon ag-hidden"
                              data-ref="eSortAsc"
                            >
                              <span
                                class="ag-icon ag-icon-asc"
                                role="presentation"
                                unselectable="on"
                              />
                            </span>
                            
        
                            <span
                              aria-hidden="true"
                              class="ag-sort-indicator-icon ag-sort-descending-icon ag-hidden"
                              data-ref="eSortDesc"
                            >
                              <span
                                class="ag-icon ag-icon-desc"
                                role="presentation"
                                unselectable="on"
                              />
                            </span>
                            
        
                            <span
                              aria-hidden="true"
                              class="ag-sort-indicator-icon ag-sort-mixed-icon ag-hidden"
                              data-ref="eSortMixed"
                            >
                              <span
                                class="ag-icon ag-icon-none"
                                role="presentation"
                                unselectable="on"
                              />
                            </span>
                            
        
                            <span
                              aria-hidden="true"
                              class="ag-sort-indicator-icon ag-sort-none-icon ag-hidden"
                              data-ref="eSortNone"
                            >
                              <span
                                class="ag-icon ag-icon-none"
                                role="presentation"
                                unselectable="on"
                              />
                            </span>
                            
    
                          </span>
                          
        
                        </div>
                        
    
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="ag-pinned-right-header ag-hidden"
              role="rowgroup"
              style="width: 0px; min-width: 0px; max-width: 0px;"
            />
          </div>
          <!-- AG Pinned Top -->
          <div
            class="ag-floating-top"
            role="presentation"
            style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
          >
            <!-- AG Row Container topLeft -->
            <div
              aria-hidden="true"
              class="ag-pinned-left-floating-top ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container topCenter -->
            <div
              class="ag-floating-top-viewport"
              role="presentation"
            >
              <div
                class="ag-floating-top-container"
                role="rowgroup"
                style="width: 200px;"
              />
            </div>
            <!-- AG Row Container topRight -->
            <div
              aria-hidden="true"
              class="ag-pinned-right-floating-top ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container topFullWidth -->
            <div
              class="ag-floating-top-full-width-container"
              role="rowgroup"
            />
          </div>
          <div
            class="ag-body ag-layout-normal"
            role="presentation"
          >
            <!-- AG Middle -->
            <div
              class="ag-body-viewport ag-row-no-animation ag-layout-normal"
              role="presentation"
            >
              <!-- AG Row Container left -->
              <div
                aria-hidden="true"
                class="ag-pinned-left-cols-container ag-hidden"
                role="rowgroup"
                style="height: 225px; width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container center -->
              <div
                class="ag-center-cols-viewport"
                role="presentation"
                style="height: 225px;"
              >
                <div
                  class="ag-center-cols-container"
                  role="rowgroup"
                  style="width: 200px; height: 225px;"
                >
                  <div
                    aria-rowindex="2"
                    class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-not-inline-editing"
                    role="row"
                    row-id="d290f1ee-6c54-4b01-90e6-d701748f0851"
                    row-index="0"
                    style="transform: translateY(0px); height: 75px;"
                    tabindex="-1"
                  >
                    <div
                      aria-colindex="1"
                      class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-column-first ag-column-last ag-cell-value"
                      col-id="fullName"
                      role="gridcell"
                      style="left: 0px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        aria-label="Joe Bloggs"
                        class="flex items-center py-2"
                        role="listitem"
                      >
                        <img
                          alt="avatar"
                          class="w-14 h-14 rounded-full mr-4"
                          src="https://i.pravatar.cc/150?img=1"
                        />
                        <div
                          class="flex flex-col"
                        >
                          <div
                            class="text-base font-medium text-[#313131]"
                          >
                            Joe Bloggs
                          </div>
                          <div
                            class="text-base font-normal text-[#475466]"
                          >
                            $11500
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="3"
                    class="ag-row-odd ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-not-inline-editing"
                    role="row"
                    row-id="f47ac10b-58cc-4372-a567-0e02b2c3d479"
                    row-index="1"
                    style="transform: translateY(75px); height: 75px;"
                    tabindex="-1"
                  >
                    <div
                      aria-colindex="1"
                      class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-column-first ag-column-last ag-cell-value"
                      col-id="fullName"
                      role="gridcell"
                      style="left: 0px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        aria-label="Jane Smith"
                        class="flex items-center py-2"
                        role="listitem"
                      >
                        <img
                          alt="avatar"
                          class="w-14 h-14 rounded-full mr-4"
                          src="https://i.pravatar.cc/150?img=2"
                        />
                        <div
                          class="flex flex-col"
                        >
                          <div
                            class="text-base font-medium text-[#313131]"
                          >
                            Jane Smith
                          </div>
                          <div
                            class="text-base font-normal text-[#475466]"
                          >
                            $36911
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="4"
                    class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-last ag-row-not-inline-editing"
                    role="row"
                    row-id="123e4567-e89b-12d3-a456-426614174000"
                    row-index="2"
                    style="transform: translateY(150px); height: 75px;"
                    tabindex="-1"
                  >
                    <div
                      aria-colindex="1"
                      class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-column-first ag-column-last ag-cell-value"
                      col-id="fullName"
                      role="gridcell"
                      style="left: 0px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        aria-label="Alice Brown"
                        class="flex items-center py-2"
                        role="listitem"
                      >
                        <img
                          alt="avatar"
                          class="w-14 h-14 rounded-full mr-4"
                          src="https://i.pravatar.cc/150?img=3"
                        />
                        <div
                          class="flex flex-col"
                        >
                          <div
                            class="text-base font-medium text-[#313131]"
                          >
                            Alice Brown
                          </div>
                          <div
                            class="text-base font-normal text-[#475466]"
                          >
                            $7200
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <!-- AG Row Container right -->
              <div
                aria-hidden="true"
                class="ag-pinned-right-cols-container ag-hidden"
                role="rowgroup"
                style="height: 225px; width: 0px; max-width: 0px; min-width: 0px;"
              />
              <!-- AG Row Container fullWidth -->
              <div
                class="ag-full-width-container"
                role="rowgroup"
                style="height: 225px;"
              />
            </div>
            <!-- AG Fake Vertical Scroll -->
            <div
              aria-hidden="true"
              class="ag-body-vertical-scroll"
            >
              
            
              <div
                class="ag-body-vertical-scroll-viewport"
                data-ref="eViewport"
              >
                
                
                <div
                  class="ag-body-vertical-scroll-container"
                  data-ref="eContainer"
                  style="height: 225px;"
                />
                
            
              </div>
              
        
            </div>
          </div>
          <!-- AG Sticky Top -->
          <div
            class="ag-sticky-top"
            role="presentation"
            style="height: 0px; top: 49px; width: 100%;"
          >
            <!-- AG Row Container stickyTopLeft -->
            <div
              aria-hidden="true"
              class="ag-pinned-left-sticky-top ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container stickyTopCenter -->
            <div
              class="ag-sticky-top-viewport"
              role="presentation"
            >
              <div
                class="ag-sticky-top-container"
                role="rowgroup"
                style="width: 200px;"
              />
            </div>
            <!-- AG Row Container stickyTopRight -->
            <div
              aria-hidden="true"
              class="ag-pinned-right-sticky-top ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container stickyTopFullWidth -->
            <div
              class="ag-sticky-top-full-width-container"
              role="rowgroup"
            />
          </div>
          <div
            class="ag-sticky-bottom ag-hidden"
            role="presentation"
            style="height: 0px; bottom: 0px; width: 100%;"
          >
            <!-- AG Row Container stickyBottomLeft -->
            <div
              aria-hidden="true"
              class="ag-pinned-left-sticky-bottom ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container stickyBottomCenter -->
            <div
              class="ag-sticky-bottom-viewport"
              role="presentation"
            >
              <div
                class="ag-sticky-bottom-container"
                role="rowgroup"
                style="width: 200px;"
              />
            </div>
            <!-- AG Row Container stickyBottomRight -->
            <div
              aria-hidden="true"
              class="ag-pinned-right-sticky-bottom ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container stickyBottomFullWidth -->
            <div
              class="ag-sticky-bottom-full-width-container"
              role="rowgroup"
            />
          </div>
          <!-- AG Pinned Bottom -->
          <div
            class="ag-floating-bottom"
            role="presentation"
            style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
          >
            <!-- AG Row Container bottomLeft -->
            <div
              aria-hidden="true"
              class="ag-pinned-left-floating-bottom ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container bottomCenter -->
            <div
              class="ag-floating-bottom-viewport"
              role="presentation"
            >
              <div
                class="ag-floating-bottom-container"
                role="rowgroup"
                style="width: 200px;"
              />
            </div>
            <!-- AG Row Container bottomRight -->
            <div
              aria-hidden="true"
              class="ag-pinned-right-floating-bottom ag-hidden"
              role="rowgroup"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            <!-- AG Row Container bottomFullWidth -->
            <div
              class="ag-floating-bottom-full-width-container"
              role="rowgroup"
            />
          </div>
          <!-- AG Fake Horizontal Scroll -->
          <div
            aria-hidden="true"
            class="ag-body-horizontal-scroll"
          >
            
            
            <div
              class="ag-horizontal-left-spacer"
              data-ref="eLeftSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
            
            <div
              class="ag-body-horizontal-scroll-viewport"
              data-ref="eViewport"
            >
              
                
              <div
                class="ag-body-horizontal-scroll-container"
                data-ref="eContainer"
                style="width: 200px;"
              />
              
            
            </div>
            
            
            <div
              class="ag-horizontal-right-spacer"
              data-ref="eRightSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
        
          </div>
          <!-- AG Overlay Wrapper -->
          <div
            class="ag-overlay ag-hidden"
            role="presentation"
          >
            
                
            <div
              class="ag-overlay-panel"
              role="presentation"
            >
              
                    
              <div
                class="ag-overlay-wrapper ag-layout-normal ag-overlay-no-rows-wrapper"
                data-ref="eOverlayWrapper"
                role="presentation"
                style="padding-top: 48px;"
              />
              
                
            </div>
            
            
          </div>
        </div>
        <div
          class="ag-tab-guard ag-tab-guard-bottom"
          role="presentation"
          tabindex="0"
        />
      </div>
      <div
        aria-hidden="true"
        class="ag-paging-panel ag-unselectable ag-focus-managed ag-hidden"
        id="ag-5"
      >
        <div
          class="ag-tab-guard ag-tab-guard-top"
          role="presentation"
          tabindex="0"
        />
        
                
        <!--AG-PAGE-SIZE-SELECTOR-->
        <span
          class="ag-paging-page-size"
          data-ref="pageSizeComp"
        />
        
                
        <span
          class="ag-paging-row-summary-panel"
        >
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            data-ref="lbFirstRowOnPage"
            id="ag-5-first-row"
          />
          
                    
          <span
            id="ag-5-to"
          >
            to
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            data-ref="lbLastRowOnPage"
            id="ag-5-last-row"
          />
          
                    
          <span
            id="ag-5-of"
          >
            of
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            data-ref="lbRecordCount"
            id="ag-5-row-count"
          />
          
                
        </span>
        
                
        <span
          class="ag-paging-page-summary-panel"
          role="presentation"
        >
          
                    
          <div
            aria-label="First Page"
            class="ag-button ag-paging-button"
            data-ref="btFirst"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-first"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Previous Page"
            class="ag-button ag-paging-button"
            data-ref="btPrevious"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-previous"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <span
            class="ag-paging-description"
          >
            
                        
            <span
              id="ag-5-start-page"
            >
              Page
            </span>
            
                        
            <span
              class="ag-paging-number"
              data-ref="lbCurrent"
              id="ag-5-start-page-number"
            />
            
                        
            <span
              id="ag-5-of-page"
            >
              of
            </span>
            
                        
            <span
              class="ag-paging-number"
              data-ref="lbTotal"
              id="ag-5-of-page-number"
            />
            
                    
          </span>
          
                    
          <div
            aria-label="Next Page"
            class="ag-button ag-paging-button"
            data-ref="btNext"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-next"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Last Page"
            class="ag-button ag-paging-button"
            data-ref="btLast"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-last"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                
        </span>
        
            
        <div
          class="ag-tab-guard ag-tab-guard-bottom"
          role="presentation"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;
